core/trino-main/src/main/java/io/trino/operator/EmptyTableFunctionPartition.java

```diff
@@ -0,0 +1,118 @@
+import io.trino.spi.type.Type;
+
+import java.util.List;
+
+import static io.airlift.concurrent.MoreFutures.toListenableFuture;
+import static io.trino.spi.StandardErrorCode.FUNCTION_IMPLEMENTATION_ERROR;
+import static io.trino.spi.ptf.TableFunctionProcessState.Type.TF_BLOCKED;
+import static io.trino.spi.ptf.TableFunctionProcessState.Type.TF_FINISHED;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static java.lang.String.format;
+
+/**
+ * This is a class representing empty input to a table function. An EmptyTableFunctionPartition is created
+ * when the table function has KEEP WHEN EMPTY property, which means that the function should be executed
+ * even if the input is empty, and all the table arguments are empty relations.
+ * <p>
+ * An EmptyTableFunctionPartition is created and processed once per node. To avoid duplicated execution,
+ * a table function having KEEP WHEN EMPTY property must have single distribution.
+ */
+public class EmptyTableFunctionPartition
+        implements TableFunctionPartition
+{
+    private final TableFunctionProcessor tableFunction;
+    private final int properChannelsCount;
+    private final int passThroughSourcesCount;
+    private final Type[] passThroughTypes;
+
+    public EmptyTableFunctionPartition(TableFunctionProcessor tableFunction, int properChannelsCount, int passThroughSourcesCount, List<Type> passThroughTypes)
+    {
+        this.tableFunction = tableFunction;
+        this.properChannelsCount = properChannelsCount;
+        this.passThroughSourcesCount = passThroughSourcesCount;
+        this.passThroughTypes = passThroughTypes.toArray(new Type[] {});
+    }
+
+    @Override
+    public WorkProcessor<Page> toOutputPages()
+    {
+        return WorkProcessor.create(() -> {
+            TableFunctionProcessState state = tableFunction.process(null);
+            if (state.type() == TF_FINISHED) {
+                return WorkProcessor.ProcessState.finished();
+            }
+            if (state.type() == TF_BLOCKED) {
+                return WorkProcessor.ProcessState.blocked(toListenableFuture(state.blockedFuture()));
+            }
+            if (state.result() != null) {
+                return WorkProcessor.ProcessState.ofResult(appendNullsForPassThroughColumns(state.result()));
+            }
+            throw new TrinoException(FUNCTION_IMPLEMENTATION_ERROR, "When function got no input, it should either produce output or return TF_BLOCKED state");
+        });
+    }
+
+    private Page appendNullsForPassThroughColumns(TableFunctionResult result)
+    {
+        Page page = result.page();
+
+        if (page.getChannelCount() != properChannelsCount + passThroughSourcesCount) {
+            throw new TrinoException(
+                    FUNCTION_IMPLEMENTATION_ERROR,
+                    format(
+                            "Table function returned a page containing %s channels. Expected channel number: %s (%s proper columns, %s pass-through index columns)",
```

Comment:
What's an "index column"?

Reply:
It is named a "pass-through index column". These are the indexes of rows, specified by the function, that should be passed to output via the pass-through mechanism.